src/consumer/fetcher.js

```diff
@@ -0,0 +1,44 @@
+const EventEmitter = require('events')
+
+/**
+ * Fetches data from all assigned nodes, waits for workerQueue to drain and repeats.
+ *
+ * @param {object} options
+ * @param {number} options.nodeId
+ * @param {import('./workerQueue').WorkerQueue} options.workerQueue
+ * @param {(nodeId: number) => Promise<T[]>} options.fetch
+ * @template T
+ */
+const createFetcher = ({ nodeId, workerQueue, fetch }) => {
+  const emitter = new EventEmitter()
+  let isRunning = false
+
+  const getWorkerQueue = () => workerQueue
+
+  const start = async () => {
+    if (isRunning) return
+    isRunning = true
+
+    while (isRunning) {
+      try {
+        const batches = await fetch(nodeId)
```

Comment:
Should we interrupt here if `!isRunning`, so that we don't push the batches to the workerQueue?

Reply:
Runner will return [] anyway when it's stopping, but I will add it here as well for clarity.